### PR TITLE
Add LibCD Support

### PR DESCRIPTION
**Pull request by LemmaEOF**
Very WIP. Will require at least one PR to RebornCore to remove the necessity of the ugly hacks I use here.

This PR adds LibCD Tweaker support for (almost) all TechReborn recipes. This means that users will be able to use JSR-223 compatible scripting languages to add recipes without writing JSON. This is immensely helpful for modpack devs, since they can programmatically create whole sets of recipes with a couple lines of code instead of having to provide large numbers of JSON files. LibCD is soft depended on using `FabricLoader.getInstance().isModLoaded()` (as LibCD doesn't have an entrypoint yet), so it will not cause crashes if not present.

Still needs finishing:
- [x] custom ingredient serializer class as to not require the hacks that `WrappedIngredient` currently uses
- [x] add Rolling Machine recipes which support [all four forms of LibCD shaped recipes](https://github.com/CottonMC/LibCD/blob/master/src/main/java/io/github/cottonmc/libcd/tweaker/RecipeTweaker.java#L168-L239)
- [x] PR to RebornCore to add ways to construct RebornRecipes other than JSON deserialization as to not require the hacks that `TweakedRebornRecipe` currently uses
- [x] add support for Fluid Generator recipes

### CODE REVIEW

**Comment by LemmaEOF** [comment]
Alright, LibCD support is fully implemented! I'm gonna write some docs and sample code soon.

**Comment by LemmaEOF** [comment]
Alright! I have the sample script fully up and running. Everything works just fine as far as I can test (#1852 and a LibCD bug notwithstanding), and every script-facing method has javadocs! I also added in a `TweakerStackGetter` for fluid cells, so a script author can get a filled cell without having to do the NBT themselves!

File: src/main/java/techreborn/compat/libcd/TRRecipeParser.java
Comment (modmuss50) [inline src/main/java/techreborn/compat/libcd/TRRecipeParser.java:89]: Quite a bit of code here that seems like it should be in libcd. Is there a specific reason why this needs to be in TR? (Thinking of the maintenance burden)

File: src/main/java/techreborn/utils/ShapedRecipeSerializer.java
Comment (modmuss50) [inline src/main/java/techreborn/utils/ShapedRecipeSerializer.java]: I will try and find some time to do this before merging this PR.

**Review by modmuss50 — COMMENTED** [review]
Looks good overall. 1 query about the code and I think cleaning up the recipe syncing is a good idea to do before pulling this in.

File: src/main/java/techreborn/compat/libcd/TRRecipeParser.java
Comment (LemmaEOF) [reply src/main/java/techreborn/compat/libcd/TRRecipeParser.java:89]: This *is* effectively in LibCD already, but TR needs its own copy because it doesn't use vanilla ingredients.

Comment (modmuss50) [reply src/main/java/techreborn/compat/libcd/TRRecipeParser.java:89]: Humm yeah, good point. What about using the wrapped indgredient? If not this isnt too much.

Comment (LemmaEOF) [reply src/main/java/techreborn/compat/libcd/TRRecipeParser.java:89]: The primary issue is that vanilla ingredients only accept one item at a time; you can't set a count like you can in TR. That's a big problem with vanilla in general, but I'm not sure how feasible it would be to have LibCD mix into all the vanilla stuff to fix it.

Comment (modmuss50) [reply src/main/java/techreborn/compat/libcd/TRRecipeParser.java:89]: Ah ok thats a good point. Dont worry about it its not like its 1000s of lines.

File: src/main/java/techreborn/utils/ShapedRecipeSerializer.java
Comment (modmuss50) [reply src/main/java/techreborn/utils/ShapedRecipeSerializer.java]: Should be sloved now, let me know if not.

**Comment by LemmaEOF** [comment]
Alright, everything looks good! This should be ready to go.

**Comment by LemmaEOF** [comment]
I *will* get rid of the ShapedRecipeSerializer quickly, though, since it's no longer needed.

**Comment by LemmaEOF** [comment]
Actually, wait. Not ready just yet; the `RollingMachine.serialize(PacketByteBuf)` and `RollingMachine.deserialize(PacketByteBuf)` need to be put somewhere else so they can be properly used. Just a minute.

**Comment by LemmaEOF** [comment]
Hmmmm. This is gonna take a bit more work than I thought. RebornRecipes themselves don't have a serialize/deserialize with a packetbytebuf, so I can either:

1. add that to RebornRecipes and give it default behavior (probably the best option)

2. add an extension of RebornRecipeType to RollingMachineRecipe

**Comment by modmuss50** [comment]
Im not sure why you need access to those? can you jump on irc as its easier to talk?

**Comment by LemmaEOF** [comment]
Turns out the issue was my local RC copy wasn't up to date. Everything's good now!

**Review by modmuss50 — APPROVED** [review]
Im happy with this now, thanks for the help.

I will merge once you give the OK :)

**Comment by LemmaEOF** [comment]
Added one last change to make the sample script log when it's run. It shouldn't be possible for the script to run if LibCD dev mode isn't enabled, but just wanna make sure. It's merge-ready!

**Comment by modmuss50** [comment]
Thanks for this :) Mind if I use the sample script on a wiki page?

**Comment by LemmaEOF** [comment]
Go ahead! That's what it's for. It's got at least one registered recipe for each TR type, along with example calls to creating fluid ingredients and using the cell stack-getter.

**Pull request merged**
